services/agent/src/simcore_service_agent/services/containers_manager.py

```diff
@@ -0,0 +1,83 @@
+import logging
+from dataclasses import dataclass, field
+
+from aiodocker import Docker
+from fastapi import FastAPI
+from models_library.api_schemas_directorv2.services import (
+    DYNAMIC_PROXY_SERVICE_PREFIX,
+    DYNAMIC_SIDECAR_SERVICE_PREFIX,
+)
+from models_library.projects_nodes_io import NodeID
+from servicelib.fastapi.app_state import SingletonInAppStateMixin
+from servicelib.utils import limited_gather
+
+from ..core.settings import ApplicationSettings
+from .docker_utils import get_containers_with_prefixes, remove_container_forcefully
+
+_logger = logging.getLogger(__name__)
+_always_visible_logger = logging.getLogger(__name__)
+
+
+@dataclass
+class ContainersManager(SingletonInAppStateMixin):
+    app_state_name: str = "containers_manager"
+
+    docker: Docker = field(default_factory=Docker)
+
+    async def force_container_cleanup(self, node_id: NodeID) -> None:
+        # compose all possible used container prefixes
+        proxy_prefix = f"{DYNAMIC_PROXY_SERVICE_PREFIX}_{node_id}"
+        dy_sidecar_prefix = f"{DYNAMIC_SIDECAR_SERVICE_PREFIX}_{node_id}"
+        user_service_prefix = f"{DYNAMIC_SIDECAR_SERVICE_PREFIX}-{node_id}"
+
+        orphan_containers = await get_containers_with_prefixes(
+            self.docker, {proxy_prefix, dy_sidecar_prefix, user_service_prefix}
+        )
+        _logger.debug(
+            "Detected orphan containers for node_id='%s': %s",
+            node_id,
+            orphan_containers,
+        )
+
+        unexpected_orphans = {
```

Comment:
Q: Is there a possibility that you catch a service that is currently being shut down as orphaned during it's shutdown? Or can this conceptually never happen? If it can happen, potentially nail down this scenario in a test.

Reply:
This is conceptually impossible.
This code is rand only after the dynamic-sidecar rand `docker compose down` and after director-v2 already removed the `proxy` and `dynamic-sidecar` services.
Call this a safety feature to cleanup if something did not function (which is the case of docker compose down that sometimes it does not work), even though we retry the command 3 times.